syntax = "proto3";

message User {
  uint32 id = 1;
  string nickname = 2;
  string mail_address = 3;
  enum UserType {
    NONE = 0;
    NORMAL = 1;
    ADMINISTRATOR = 2;
    GUEST = 3;
    DISABLED = 4;
  }
  UserType user_type = 4;
}

message UserRequest {
  uint32 id = 1;
}

message UserTypeRequest {
  enum UserType {
    NONE = 0;
    NORMAL = 1;
    ADMINISTRATOR = 2;
    GUEST = 3;
    DISABLED = 4;
  }
  UserType user_type = 4;
}

message UserResponse {
  bool error = 1;
  string message = 2;
  User user = 3;
}

message UserCntResponse {
  bool error = 1;
  uint32 user_cnt = 2;
}

service UserManager {
  rpc getUser (UserRequest) returns (UserResponse) {}
  rpc addUser (User) returns (UserResponse) {}
  rpc countAlreadyUsers (stream UserRequest) returns (UserCntResponse) {}
  rpc getUsersByType (UserTypeRequest) returns (stream UserResponse) {}
  rpc getUsersByIds (stream UserRequest) returns (stream UserResponse) {}
}
